faust/stores/bigtable.py

```diff
@@ -0,0 +1,605 @@
+        return key
+
+    def _get_partition_from_bigtable_key(self, key: bytes) -> int:
+        separator = b"_..._"
+        partition_str, _ = key.rsplit(separator, 1)
+        return int(partition_str)
+
+    def _active_partitions(self) -> List[int]:
+        actives = self.app.assignor.assigned_actives()
+        topic = self.table.changelog_topic_name
+        partitions = []
+        for partition in range(self.app.conf.topic_partitions):
+            tp = TP(topic=topic, partition=partition)
+            if tp in actives or self.table.is_global:
+                partitions.append(partition)
+        return partitions
+
+    def _get_current_partitions(self) -> List[int]:
+        event = current_event()
+        if (
+            event is not None
+            and event.message.topic is not None
+            and not self.table.is_global
+            and not self.table.use_partitioner
+        ):
+            partition = event.message.partition
+            return [partition]
+        return self._active_partitions()
+
+    def _get_partitions_for_key(self, key: bytes) -> List[int]:
+        if key in self._key_index:
+            return [self._key_index[key]]
+        return self._get_current_partitions()
+
+    @staticmethod
+    def bigtable_exrtact_row_data(row_data):
+        return list(row_data.to_dict().values())[0][0].value
+
+    def _del_cache(self, key: bytes):
+        if self._startup_cache is not None:
+            self._startup_cache[key] = None
+
+    def _set_cache(self, key: bytes, value):
+        if self._startup_cache is not None:
+            self._startup_cache[key] = value
+
+    def _get_cache(self, key: bytes):
+        if self._startup_cache_enable and self._startup_cache is not None:
+            if key in self._startup_cache:
+                return self._startup_cache[key], True
+        return None, False
+
+    def _invalidate_startup_cache(self):
+        if self._startup_cache is not None:
+            self._startup_cache.clear()
+            self._startup_cache_partitions = set()
+            gc.collect()
+            self.log.info(f"Invalidated startup cache for table {self.table_name}")
+        self._invalidation_timer.cancel()
+        del self._invalidation_timer
+        self._invalidation_timer = None
+
+    def _set_mutation(self, mutated_row: DirectRow):
+        self._mutation_batcher.mutate(mutated_row)
+
+    def _bigtable_get(self, keys: List[bytes]) -> Tuple[Optional[bytes], Optional[int]]:
+        rowset = BT.RowSet()
+        for key in keys:
+            rowset.add_row_key(key)
+        if self._mutation_batcher_enable:
+            self._mutation_batcher.flush()
+
+        rows = self.bt_table.read_rows(row_set=rowset, filter_=self.row_filter)
+        for row in rows:
+            if row is not None:
+                partition = self._get_partition_from_bigtable_key(row.row_key)
+                return self.bigtable_exrtact_row_data(row), partition
```

Comment:
This return seems wrong to me. If we have multiple rows we stop after return the first match.

Reply:
This is taken from rocksdb.py 

```python
...
        for partition, db in dbs:
            if db.key_may_exist(key)[0]:
                value = db.get(key)
                if value is not None:
                    self._key_index[key] = partition
                    return _DBValueTuple(db, value)
...
```